### ISSUE TITLE
Certain posts cause a 500 Internal Error

### ISSUE THREAD

**Issue by thor-the-norseman** [tootsuite/mastodon#1958]
Sometimes, it's impossible to post things. It's been a problem for a while. In this particular case, a reply to a thread:

```
curl 'https://snabeltann.no/api/v1/statuses' \
-XPOST \
-H 'Content-Type: application/json;charset=UTF-8' \
-H 'Referer: https://snabeltann.no/web/statuses/24420' \
-H 'Authorization: Bearer CENSORED' \
-H 'Accept: application/json, text/plain, */*' \
-H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_4) AppleWebKit/603.1.30 (KHTML, like Gecko) Version/10.1 Safari/603.1.30' \
-H 'Origin: https://snabeltann.no' \
--data-binary '{"status":"@ashleykrista@mastodon.technology Hahaha! \u0010Sounds good.\n\nMore facts about Norwegians:\n\nhttps://www.youtube.com/watch?v=hGOLJlZW_qc\n\nN.I.S.S.E. (B.R.O.W.N.I.E.) - Norwegian Inherited Stiffness SyndromE","in_reply_to_id":24420,"media_ids":[],"sensitive":false,"spoiler_text":"","visibility":"unlisted"}'```
```

For some reason, on Safari on macOS Sierra, the exclamation point and whitespace after "Hahaha" gets converted to "! \u0010".

This causes the following error on the web process:

```
[ac49fff1-50d8-49e7-9858-e949f27bf8da] method=POST path=/api/v1/statuses format=html controller=Api::V1::StatusesController action=create status=500 error='SyntaxError: '\#x10' is not a valid XML character.' duration=135.69 view=0.00 db=26.36
[ac49fff1-50d8-49e7-9858-e949f27bf8da]   
[ac49fff1-50d8-49e7-9858-e949f27bf8da] SyntaxError ('\#x10' is not a valid XML character.):
[ac49fff1-50d8-49e7-9858-e949f27bf8da]   
[ac49fff1-50d8-49e7-9858-e949f27bf8da] app/lib/atom_serializer.rb:10:in `dump'
[ac49fff1-50d8-49e7-9858-e949f27bf8da] app/lib/atom_serializer.rb:10:in `render'
[ac49fff1-50d8-49e7-9858-e949f27bf8da] app/services/concerns/stream_entry_renderer.rb:5:in `stream_entry_to_xml'
[ac49fff1-50d8-49e7-9858-e949f27bf8da] app/services/process_mentions_service.rb:36:in `block in call'
[ac49fff1-50d8-49e7-9858-e949f27bf8da] app/services/process_mentions_service.rb:30:in `call'
[ac49fff1-50d8-49e7-9858-e949f27bf8da] app/services/post_status_service.rb:25:in `call'
[ac49fff1-50d8-49e7-9858-e949f27bf8da] app/controllers/api/v1/statuses_controller.rb:60:in `create'
[ac49fff1-50d8-49e7-9858-e949f27bf8da] app/controllers/concerns/localized.rb:24:in `block in set_locale'
[ac49fff1-50d8-49e7-9858-e949f27bf8da] app/controllers/concerns/localized.rb:23:in `set_locale'
```

When I remove the exclamation point, the post goes through.

I'm sadly unable to reproduce the exact conditions. Other messages with an exclamation point, the word "Hahaha" or a URL did not produce an internal error. I'll try to create a consistently reproducible case.

**Comment by thor-the-norseman** [comment]
If I'm not mistaken, \u0010 is equivalent to ASCII character 16. But it's interesting to note that the decimal notation of \n would be 0010 if you padded it with two zeroes.

**Comment by thor-the-norseman** [comment]
Oh, and it seems that the posts were somehow created anyway, but are impossible to delete. This is the thread in question:

The DELETE request looks like:

```
curl 'https://snabeltann.no/api/v1/statuses/24434' \
-XDELETE \
-H 'Referer: https://snabeltann.no/web/statuses/24420' \
-H 'Authorization: Bearer 493d0a2b02601eed69d3c46578f87aae9a72073e09548fd1886ae85be15d887c' \
-H 'Accept: application/json, text/plain, */*' \
-H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_4) AppleWebKit/603.1.30 (KHTML, like Gecko) Version/10.1 Safari/603.1.30' \
-H 'Origin: https://snabeltann.no'
```

**Comment by happycoloredbanana** [comment]
(no connection to the issue)
@thor-the-norseman, you are posting real OAuth tokens, probably not the best idea.

**Comment by thor-the-norseman** [comment]
Removing them.

**Comment by happycoloredbanana** [comment]
Maybe connected to #1870

**Comment by wxcafe** [comment]
Probably related to #1870 yeah.

**Comment by ghostfromicosahedron** [comment]
I started experiencing this about 24 hours ago... and I don't know if it helps to know when I've experienced it:

At first the issue was almost exclusively when I'm @ replying to something, or trying to post anything with the @ symbol. However, occasionally (VERY occasionally), the reply will post. But a few hours later (and as of this posting), I'd estimate 9 out of every 10 attempts to post any content at all will fail. Posts are more likely to fail if they contain some kind of punctuation or symbol (.,'/?@#$), but even then, it's really not predictable at all. Sometimes just removing a question mark will let the post go through.

The error message says "Failed to post status with error: (null)"

I'm using Amaroq, iOS 9.3.5. It ONLY happens on toot.cat, and has not happened on several other instances I've tried so far.

**Comment by thor-the-norseman** [comment]
@ghostfromicosahedron I think you're talking about a different issue. What I'm talking about is much older than 24 hours, and is happening on my own instance, snabeltann.no, using the official web app. The issue is with the web app code that prepares the POST request, I suspect, together with broken sanitizer code on the server. Amaroq doesn't even use code from the web app, so it can't possibly have the same exact bug. It could be breaking the sanitizer in some other way, of course, but again, I think that's a different bug...

**Comment by thor-the-norseman** [comment]
...and more crucially, problems in Amaroq should be reported to Amaroq's creator. He has to check his code for problems before Mastodon gets involved.

**Comment by ghostfromicosahedron** [comment]
it was. and he did. and that's why i'm posting it here.

if it's not related to this bug, then just disregard my comment. it just seemed real similar. no big deal.

**Comment by MightyPork** [comment]
might have come across this in my accidental fuzzing toot here -> https://github.com/tootsuite/mastodon/issues/2728

**Comment by Gargron** [comment]
Has this been fixed just like #1870?
